Follow up to implementation of window.rs module. Also adds a new `Float` trait for common generic operations. [Minor Breaking Change]

mitchmindtree opened the pull request:
@andrewcsmith I left some more detailed notes in the commit messages - let me know your thoughts on the changes to the `window.rs` module :+1:

Follow up to #36

mitchmindtree commented:
I just put the new `window.rs` module behind a "window" feature. This way the crate can remain minimal while offering users the ability to opt-in to the new windowing functionality via `cargo build --features="window"` or adding the feature to their projects' cargo.toml file. Thoughts?

andrewcsmith commented on src/window.rs:
```diff
@@ -1,73 +1,93 @@
     pub bin: usize,
+    /// The step size over `frames` for the start of each new `Windowed` chunk.
     pub hop: usize,
-    pub idx: usize,
```
Why does the idx and data need to be private? Either should be able to be changed without any side effects.

mitchmindtree replied:
Hmmm, my thoughts were that making them private might ensure a little more invariance in the behaviour of the `Windower` (I actually meant to make `bin` and `hop` private too), and that it's almost as easy/efficient to construct a new `Windower` as it is to mutate the fields anyway. Now that you mention it though, I can't see much harm being done in having them public, and it seems like you might have  a particular use-case in mind so I'm happy to re-add them!

andrewcsmith commented:
Looks excellent. I'm not quite sure why this actually needs to be feature-gated, cause I can indeed see a few quite simple time-domain use-cases (like generating grains for granular synthesis, as I mentioned) that don't require holding onto slices of data. But either way is fine, obviously.

I'm going to refactor [vox_box.rs](https://github.com/andrewcsmith/vox_box.rs) to work with `sample` in the next couple weeks. That should be a good test-case, and perhaps some of those elements can make their way into `sample`. (Those would definitely have to be feature-gated, since they mostly depend on holding slices of frames in memory.)

mitchmindtree commented:
@andrewcsmith thanks!

I think you're right, I was getting a bit eager with the feature gating, I've removed the window feature in favour of waiting until it's actually needed/requested.

Also, while addressing [your comment](https://github.com/RustAudio/sample/pull/38#discussion_r66284667) I realised that we can trim down the `Windower` a little by removing the `idx` field in favour of re-slicing the frame slice from the next hop position. I've also made the `frames` field public and fixed the Windower::size_hint implementation. If you're happy with this, I'll merge and publish. I'm looking forward to hearing how sample goes in vox_box btw!

andrewcsmith commented:
Excellent. Happy to see this merged. Thanks for the excellent feedback, all around.

mitchmindtree commented:
Published as **0.6.1**.
